src/table/operations.jl

```diff
@@ -1,63 +1,127 @@
+DTable with 2 partitions
+Tabletype: NamedTuple
+
+julia> fetch(m)
+(r = [2, 3, 4],)
+
+julia> m = map(x -> (r1 = x.a + x.b, r2 = x.a - x.b), d)
+DTable with 2 partitions
+Tabletype: NamedTuple
+
+julia> fetch(m)
+(r1 = [2, 3, 4], r2 = [0, 1, 2])
 ```
 """
-function map(f, d::DTableRows)
-    chunk_wrap = (chunk, f) -> begin
-        res = chunk |> TableOperations.map(x-> (result=f(x),)) |> Tables.materializer(chunk)
-        Tables.getcolumn(res, :result)
+function map(f, d::DTable)
+    chunk_wrap = (_chunk, _f) -> begin
+        if isvalid(_chunk)
+            m = TableOperations.map(_f, _chunk)
+            Tables.materializer(_chunk)(m)
+        else
+            _chunk
+        end
     end
-    result_vector = map(c -> Dagger.spawn(chunk_wrap, c, f), d.dtable.chunks)
-    Dagger.@spawn vcat(result_vector...)
+    chunks = map(c -> Dagger.spawn(chunk_wrap, c, f), d.chunks)
+    DTable(chunks, d.tabletype)
 end
 
 """
-    reduce(f_reduce::Function, f_row::Function, d::DTableRows; init)
+    reduce(f, d::DTable; cols=nothing, [init]) -> NamedTuple
+
+Reduces `d` using function `f` applied on all columns of the DTable.
+
+By providing the kwarg `cols` as a `Vector{Symbol}` object it's possible 
+to restrict the reduction to the specified columns.
+The reduced values are provided in a NamedTuple under names of reduced columns.
 
-Reduces `d` using function `f_reduce` on values obtained from applying `f_row` on its rows.
+For the `init` kwarg please refer to `Base.reduce` documentation,
+as it follows the same principles. 
 
 # Examples
 ```
-    d = DTable((a=rand(1000), b=rand(1000)), chunksize=100)
-    r = reduce(+, row -> row.a + row.b, eachrow(d); init=0.0)
-    fetch(r)
+julia> d = DTable((a = [1, 2, 3], b = [1, 1, 1]), 2);
+
+julia> r1 = reduce(+, d)
+EagerThunk (running)
+
+julia> fetch(r1)
+(a = 6, b = 3)
+
+julia> r2 = reduce(*, d, cols=[:a])
+EagerThunk (running)
+
+julia> fetch(r2)
+(a = 6,)
 ```
 """
-function reduce(f_reduce::Function, f_row::Function, d::DTableRows; init)
-    chunk_results = map(c -> Dagger.spawn(_chunk_reduce, f_reduce, f_row, c, init), d.dtable.chunks)
-    reduce_chunks = (f, _init, c_res...) -> reduce(f, c_res; init=_init)
-    Dagger.@spawn reduce_chunks(f_reduce, init, chunk_results...)
-end
+function reduce(f, d::DTable; cols=nothing::Union{Nothing, Vector{Symbol}}, init=Base._InitialValue())
+    # TODO replace this with checking the colnames in schema, once schema handling gets introduced
+    if length(d.chunks) > 0
+        columns = isnothing(cols) ? Tables.columnnames(Tables.columns(_retrieve(d.chunks[1]))) : cols
+    else
+        return Dagger.@spawn NamedTuple()
+    end
 
-function _chunk_reduce(f_reduce, f_row, chunk, init)
-    mapped = chunk |> TableOperations.map(x -> (result = f_row(x)))
-    reduce(f_reduce, mapped; init=init)
+    chunk_reduce = (_f, _chunk, _cols, _init) -> begin
+        values = [reduce(_f, Tables.getcolumn(_chunk, c); init=deepcopy(_init)) for c in _cols]
```

Comment:
@jpsamaroo Do you think the per-column-within-chunk reduction should be done within separate tasks as well? 
I think this could be a potential performance improvement with bigger chunks

Reply:
We should definitely parallelize more rather than less, and parallelizing per-column within a chunk might give the scheduler better information about compute costs for reducing that column, since columns can have different types, and thus take more or less time to compute.

Reply:
I've added the spawns per column inside and:

1. Performance wise probably only an upgrade for many columns (need to look for a threshold sometime). For 2-4 cols it was usually a downgrade.
2. Stability wise it causes this https://github.com/JuliaParallel/Dagger.jl/issues/267 , so for now I'll keep this commented out

```julia
    col_in_chunk_reduce = (_f, _c, _init, _chunk) -> reduce(_f, Tables.getcolumn(_chunk, _c); init=deepcopy(_init))

    chunk_reduce = (_f, _chunk, _cols, _init) -> begin
        if length(_cols) <= 1 
            v = [col_in_chunk_reduce(_f, c, _init, _chunk) for c in _cols]
        else
            values = [Dagger.spawn(col_in_chunk_reduce, _f, c, _init, _chunk) for c in _cols]
            v = fetch.(values)
        end
        (; zip(_cols, v)...)
    end
```